src/ctap/data_formats.rs

```diff
@@ -447,56 +450,72 @@ pub struct PublicKeyCredentialSource {
     pub cred_random: Option<Vec<u8>>,
 }
 
+// We serialize credentials for the persistent storage using CBOR maps. Each field of a credential
+// is associated with a unique tag, implemented with a CBOR unsigned key.
+enum PublicKeyCredentialSourceField {
+    CredentialId = 0,
+    PrivateKey = 1,
+    RpId = 2,
+    UserHandle = 3,
+    OtherUi = 4,
+    CredRandom = 5,
+    // When a field is removed, its tag should be reserved and not used for new fields. We document
+    // those reserved tags below.
+    // Reserved tags: none.
+}
+
+impl From<PublicKeyCredentialSourceField> for cbor::KeyType {
+    fn from(field: PublicKeyCredentialSourceField) -> cbor::KeyType {
+        (field as u64).into()
+    }
+}
+
 impl From<PublicKeyCredentialSource> for cbor::Value {
     fn from(credential: PublicKeyCredentialSource) -> cbor::Value {
-        let mut private_key = [0u8; 32];
+        use PublicKeyCredentialSourceField::*;
+        let mut private_key = [0; 32];
         credential.private_key.to_bytes(&mut private_key);
-        let other_ui = match credential.other_ui {
-            None => cbor_null!(),
-            Some(other_ui) => cbor_text!(other_ui),
-        };
-        let cred_random = match credential.cred_random {
-            None => cbor_null!(),
-            Some(cred_random) => cbor_bytes!(cred_random),
-        };
-        cbor_array! {
-            credential.credential_id,
-            private_key,
-            credential.rp_id,
-            credential.user_handle,
-            other_ui,
-            cred_random,
+        cbor_map_options! {
+            CredentialId => Some(credential.credential_id),
+            PrivateKey => Some(private_key.to_vec()),
+            RpId => Some(credential.rp_id),
+            UserHandle => Some(credential.user_handle),
+            OtherUi => credential.other_ui,
+            CredRandom => credential.cred_random
         }
     }
 }
 
-impl TryFrom<cbor::Value> for PublicKeyCredentialSource {
+impl TryFrom<&cbor::Value> for PublicKeyCredentialSource {
```

Comment:
What is the rationale for taking the CBOR value by reference? In principle, the CBOR value is a short-lived object, out of which we could move elements in this `try_from`. All the other types parse it by value.

Reply:
I was anticipating the "consistency" argument to avoid too many ping-pong since all other types parse a `cbor::Value` by reference (I don't know where you saw that they do it by value). I'll be happy to change it back to call-by-value if you think it's fine to not be "consistent".

Reply:
The reason is that most other data structures appear as values in maps. We then get references of those maps entries. This data structure is not used as a parameter, but for our own sake. So I don't mind the "inconsistency" here.

Reply:
Done.

Note that _the fact that the other data structures appear as values in maps_ is not the real reason why it's not call-by-value. It's possible to extract the value from a map (as done for the credential source in the latest commit). If it's not possible to use call-by-value, it would be because the initial CBOR value is used more than once (breaking the linearity required by call-by-value). I don't know if this is the case. If it's not the case, then the other functions could as well use call-by-value and avoid unnecessary allocations. But that would be another PR and only useful if we believe allocations are an issue.